pulsar-broker/src/main/java/org/apache/pulsar/broker/service/InMemoryAndPreventCycleFilterRedeliveryTracker.java

```diff
@@ -0,0 +1,173 @@
+        if (position == null || consumer == null){
+            return newCount;
+        }
+        Position originalEarliestPosition = redeliveryStartAt;
+        Position actualEarliestPosition = null;
+        if (originalEarliestPosition == null) {
+            cleanEarliestInformation();
+            actualEarliestPosition = position;
+        } else {
+            int isEarlier = comparePosition(originalEarliestPosition, position);
+            if (isEarlier < 0) {
+                return newCount;
+            } else if (isEarlier > 0) {
+                cleanEarliestInformation();
+                actualEarliestPosition = position;
+            } else {
+                actualEarliestPosition = originalEarliestPosition;
+            }
+        }
+        redeliveryStartAt = actualEarliestPosition;
+        int redeliveryCount = earliestEntryRedeliveryCountMapping.computeIfAbsent(consumer, c -> new AtomicInteger())
+                .incrementAndGet();
+        if (redeliveryCount >= 3) {
+            pausedConsumers.put(consumer, new PauseConsumerInformation(actualEarliestPosition));
+        } else {
+        }
+        return newCount;
+    }
+
+    @Override
+    public void remove(Position position, Position markDeletedPosition) {
+        super.remove(position, markDeletedPosition);
+        if (redeliveryStartAt == null || comparePosition(redeliveryStartAt, position) == 0
+                || comparePosition(redeliveryStartAt, markDeletedPosition) >= 0) {
+            cleanEarliestInformation();
+        }
+    }
+
+    @Override
+    public void clear() {
+        super.clear();
+        cleanEarliestInformation();
+    }
+
+    public boolean isConsumerPaused(Consumer consumer) {
+        if (consumer == null) {
+            return false;
+        }
+        PauseConsumerInformation pauseConsumerInformation = pausedConsumers.get(consumer);
+        if (pauseConsumerInformation == null) {
+            return false;
+        }
+        if (!pauseConsumerInformation.isValid(redeliveryStartAt)) {
+            pausedConsumers.remove(consumer);
+            return false;
+        }
+        return true;
+    }
+
+    public int pausedConsumerCount() {
+        return (int) pausedConsumers.keySet().stream().map(this::isConsumerPaused).count();
+    }
+
+    private void cleanEarliestInformation() {
+        redeliveryStartAt = null;
+        // If consumer has been closed, remove this consumer.
+        List<Consumer> closedConsumers = earliestEntryRedeliveryCountMapping.keySet().stream()
+                .filter(Consumer::isClosed).toList();
+        closedConsumers.forEach(earliestEntryRedeliveryCountMapping::remove);
+        // Just reset counter to 0, because value will be removed if consumer is closed.
+        earliestEntryRedeliveryCountMapping.values().forEach(i -> i.set(0));
+        pausedConsumers.clear();
+    }
+
+    private static int comparePosition(Position pos1, Position pos2) {
```

Comment:
this method should go somewhere in Position or in some utility class
like comparePositionByEntryId

please check in the code if we already have something like that

Reply:
Already fixed. I didn't find the right tool method, so I added a new one.

Reply:
it seems you are not using the new utility method you introduce here

Reply:
Yes

Reply:
this method is not used, we should drop it

Reply:
already removed